### PR TITLE
Introduce a kani macro for checking coverage

### LINKED ISSUE #696: Create a `cover` statement to RMC
**Requested feature:** Provide a functionality to RMC to allow users to prove reachability. I.e., prove that there is at least one path where the provided statement is true. 
**Use case:** Provides more insight to proof harnesses allowing users to ensure that their harnesses are not over-constrained.
**Link to relevant documentation (Rust reference, Nomicon, RFC):**
**Is this a breaking change?** No

**Test case:**

```rust

#[proof]
pub fn check_update() {
    // Write any value to the database.
    let database = rmc::nondet::<MyDatabase>();
    let key = rmc::nondet::<Key>();
    let v1 = rmc::nondet::<Value>();
    database.write(key, v1);

    // Write any other value to the database under the same key.
    let v2 = rmc::nondet::<Value>();
    database.write(key, v2);

    // Make sure v1 may be different than v2. It' OK if it's equal though.
    // Proof should fail if this is always false though.
    rmc::cover(v1 != v2);

    // Verify read gets the latest value.
    assert!(database.read(key) == v2);
}
```

This dummy harness shows how the harness runs under every single value combination, and that the last read is reachable under scenarios where v1 is different than v2. This is different than using an `assume` or another `assert` statement, since it won't restrict the search space.

Note that the verification should fail if the `cover` statement is unreachable.

### PR DESCRIPTION
**Pull request by zhassan-aws**
### Description of changes: 

Introduce a `kani::cover` macro that can be used for checking whether certain conditions can occur without causing verification to fail.

### Resolved issues:

Resolves #696 

### Related RFC:

Optional #1905 

### Call-outs:

This PR does not yet introduce reachability checks for cover properties. Thus, both unreachable and reachable but unsatisfiable properties will be reported as "UNSATISFIABLE". I'm planning to introduce reachability checks in a follow-up PR because I didn't want to complicate this one too much.

### Testing:

* How is this change tested? Added a few tests.

* Is this a refactor change? No

### Checklist
- [X] Each commit message has a non-empty body, explaining why the change was made
- [X] Methods or procedures are documented
- [X] Regression or unit tests are included, or existing tests cover the modified code
- [X] My PR is restricted to a single feature or bugfix

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 and MIT licenses.

### CODE REVIEW

**Comment by tedinski** [comment]
Hmm, was this always going to be a macro, not just a function, and I missed that when I glanced at the RFC before?

Why a macro over just a function? Is it just the messages part? Is that essential?

I want to make sure the disadvantages of a macro (complexity, compile time, IDE autocomplete problems) are actually needed for this...

**Comment by zhassan-aws** [comment]
> Hmm, was this always going to be a macro, not just a function, and I missed that when I glanced at the RFC before?

No, it was originally a function, but the discussion on the RFC led to introducing a macro. See https://github.com/model-checking/kani/pull/1906#discussion_r1023433867.

The function still exists though, and is public, so can be used directly.

File: kani-driver/src/cbmc_property_renderer.rs [hunk 1]
```diff
@@ -240,6 +240,12 @@ pub fn format_result(properties: &Vec<Property>, show_checks: bool) -> String {
     let mut number_tests_undetermined = 0;
```
Comment (celinval) [inline kani-driver/src/cbmc_property_renderer.rs:240]: I wonder if we should rename this to `number_checks_*` to contrast with `number_covers_*`.

File: kani-driver/src/cbmc_property_renderer.rs [hunk 2]
```diff
@@ -240,6 +240,12 @@ pub fn format_result(properties: &Vec<Property>, show_checks: bool) -> String {
...
     let mut failed_tests: Vec<&Property> = vec![];
 
+    // cover checks
```
Comment (celinval) [inline kani-driver/src/cbmc_property_renderer.rs:243]: Can you please add to the function documentation that we separate these two in the report?

File: library/kani/src/lib.rs [hunk 3]
```diff
@@ -64,6 +64,17 @@ pub const fn assert(_cond: bool, _msg: &'static str) {
     }
 }
 
+/// Creates a cover property with the specified condition and message.
```
Comment (celinval) [inline library/kani/src/lib.rs:67]: Can you please expand a bit on what a cover property is?

File: tests/expected/cover/cover-pass/expected [hunk 4]
```diff
@@ -0,0 +1,11 @@
+Status: SATISFIED\
+Description: "cover condition: sorted == arr"\
+main.rs:12:5 in function cover_sorted
+
+Status: SATISFIED\
+Description: "cover condition: sorted != arr"\
+main.rs:13:5 in function cover_sorted
+
+ ** 0 of 2 cover properties unsatisfiable
```
Comment (celinval) [inline tests/expected/cover/cover-pass/expected:9]: Can we invert the report? I think we should report the number of covers that were satisfied.

**Review by celinval — COMMENTED** [review]
Thanks for doing this!

File: kani-compiler/src/codegen_cprover_gotoc/overrides/hooks.rs [hunk 5]
```diff
@@ -46,6 +46,39 @@ fn matches_function(tcx: TyCtxt, instance: Instance, attr_name: &str) -> bool {
     false
 }
 
+struct Cover;
+impl<'tcx> GotocHook<'tcx> for Cover {
```
[the remark this answers is not in the document]
Comment (zhassan-aws) [reply kani-compiler/src/codegen_cprover_gotoc/overrides/hooks.rs:58]: Added.

File: kani-driver/src/cbmc_property_renderer.rs (hunk 2 above)
Comment (zhassan-aws) [reply kani-driver/src/cbmc_property_renderer.rs:248]: Done.

File: tests/expected/cover/cover-pass/expected (hunk 4 above)
Comment (zhassan-aws) [reply tests/expected/cover/cover-pass/expected]: I agree an inverted report is better (I made the change). The reason I didn't do it initially, is that an inverted report is a bit inconsistent with the report for normal checks where we reported the number of things that went wrong (failures) as opposed to those that were successful (similarly, the unsatisfiable covers are the ones that are unexpected).

I'm pro inverting the summary for normal checks as well (perhaps in a separate PR).

File: library/kani/src/lib.rs (hunk 3 above)
Comment (zhassan-aws) [reply library/kani/src/lib.rs:67]: Done

File: kani-driver/src/cbmc_property_renderer.rs (hunk 1 above)
Comment (zhassan-aws) [reply kani-driver/src/cbmc_property_renderer.rs]: Done.

**Review by celinval — APPROVED** [review]
Thanks!

**Pull request merged**